source/bson-corpus/tests/document.json

```diff
@@ -17,6 +17,26 @@
             "description": "Single-character key subdoc",
             "canonical_bson": "160000000378000E0000000261000200000062000000",
             "canonical_extjson": "{\"x\" : {\"a\" : \"b\"}}"
+        },
+        {
+            "description": "Dollar-prefixed key in sub-document",
+            "canonical_bson": "170000000378000F000000022461000200000062000000",
+            "canonical_extjson": "{\"x\" : {\"$a\" : \"b\"}}"
+        },
+        {
+            "description": "Dollar as key in sub-document",
+            "canonical_bson": "160000000378000E0000000224000200000061000000",
+            "canonical_extjson": "{\"x\" : {\"$\" : \"a\"}}"
```

Comment:
Curious here and elsewhere, do we also need to test allowing dots and dollars in sub-arrays? Some drivers may have different logical paths for parsing sub-arrays vs. embedded documents.

Reply:
Do you mean as the field in a BSON array itself, or an embedded document within a BSON array?

In the case of the field in a BSON array, those are typically assigned as numeric strings starting at `"0"`, but there is really no enforcement of that. When iterating a BSON array (i.e. type 0x04), the keys should be ignored. Such a test would likely require we use `degenerate_bson` (e.g. ["Single Element Array with index set incorrectly to ab"](https://github.com/mongodb/specifications/blob/master/source/bson-corpus/tests/array.json#L23)).

As for an embedded document _within_ a BSON array, I'm not sure why any driver would have a different code path for that vs. an embedded document anywhere else. If anyone is aware of such a code path, I'm happy to add an extra test case; otherwise, I don't think it's necessary.

Reply:
That's true; it would have to be an embedded document within a BSON array, and drivers shouldn't have different paths for those vs. regular embedded documents. Probably not necessary, then.